File: frontend/src/services/menuService.ts
import api from './api';
import { getDemoVendorMenu, isDemoVendorId } from '../data/demoVendors';
import { MenuItem } from '../types';

export interface DishSearchResult {
  vendor: {
    id: string;
    name: string;
    imageUrl?: string;
    rating: number;
    totalRatings: number;
  };
  items: MenuItem[];
}

export async function getVendorMenu(vendorId: string): Promise<MenuItem[]> {
  if (isDemoVendorId(vendorId)) {
    return getDemoVendorMenu(vendorId);
  }

  const response = await api.get<MenuItem[]>(`/vendors/${vendorId}/menu`);
  return response.data;
}

export async function searchMenuItems(
  query: string,
  filters?: { category?: string; priceMin?: number; priceMax?: number },
): Promise<DishSearchResult[]> {
  try {
    const params = new URLSearchParams({ q: query });
    if (filters?.category) params.set('category', filters.category);
    if (filters?.priceMin !== undefined) params.set('priceMin', String(filters.priceMin));
    if (filters?.priceMax !== undefined) params.set('priceMax', String(filters.priceMax));

    const response = await api.get<DishSearchResult[]>(`/menu/search?${params}`);
    return response.data;
  } catch {
    return [];
  }
}

export async function createMenuItem(payload: Partial<MenuItem>): Promise<MenuItem> {
  const response = await api.post<MenuItem>('/menu', payload);
  return response.data;
}

export async function updateMenuItem(
  id: string,
  payload: Partial<MenuItem>,
): Promise<MenuItem> {
  const response = await api.put<MenuItem>(`/menu/${id}`, payload);
  return response.data;
}

export async function deleteMenuItem(id: string): Promise<{ success: boolean; removed?: boolean }> {
  const response = await api.delete<{ success: boolean; removed?: boolean }>(`/menu/${id}`);
  return response.data;
}

File: flutter_geobite/lib/services/menu_service.dart
import '../models/menu_item.dart';
import '../core/api_client.dart';

class MenuService {
  Future<List<MenuItem>> getVendorMenu(String vendorId) async {
    try {
      final response = await apiClient.dio.get('/vendors/$vendorId/menu');
      final data = response.data;
      final List list = data is Map ? (data['data'] ?? []) : data;
      return list.map((m) => MenuItem.fromJson(m)).toList();
    } catch (e) {
      throw Exception('Failed to load menu: $e');
    }
  }

  // Helper method for the UI
  Future<MenuItem> addMenuItem(MenuItem item) async {
    return createMenuItem({
      'vendorId': item.vendorId,
      'name': item.name,
      'description': item.description,
      'price': item.price,
      'isAvailable': item.isAvailable,
      'imageUrl': item.imageUrl,
    });
  }

  Future<MenuItem> createMenuItem(Map<String, dynamic> payload) async {
    try {
      final response = await apiClient.dio.post('/menu', data: payload);
      return MenuItem.fromJson(response.data);
    } catch (e) {
      throw Exception('Failed to add menu item: $e');
    }
  }

  Future<MenuItem> updateMenuItem(String id, Map<String, dynamic> payload) async {
    try {
      // NestJS conventions usually prefer PATCH for partial updates
      final response = await apiClient.dio.patch('/menu/$id', data: payload);
      return MenuItem.fromJson(response.data);
    } catch (e) {
      throw Exception('Failed to update menu item: $e');
    }
  }

  Future<bool> deleteMenuItem(String id) async {
    try {
      final response = await apiClient.dio.delete('/menu/$id');
      return response.statusCode == 200 || response.statusCode == 204;
    } catch (e) {
      throw Exception('Failed to delete menu item: $e');
    }
  }
}

final menuService = MenuService();
